Handle normalization of unicode identifiers

da-woods opened the pull request:
Also add a test-case. (Also add a test case for passing keyword
arguments, which already worked, but is good to have as a test
case).

PEP3131 implies that normalization of identifiers should be done before tokenization (i.e. the moment any unicode character is found the surrounding area of code should be normalized then checked to see if it's an identifier). Practically, I don't think this is what Python does. Instead it finds a bit of string that might be an identifier and validates that it is (https://github.com/python/cpython/blob/69f37bcb28d7cd78255828029f895958b5baf6ff/Parser/tokenizer.c#L1062). Normalization is done when the AST is generated (https://github.com/python/cpython/blob/51aac15f6d525595e200e3580409c4b8656e8a96/Python/ast.c#L610).

I've followed (what I believe is) Python's approach. This might make a different if some unicode characters that aren't valid start or continuation characters can get normalized to something that is. I don't think this happens though. And this patch should certainly take care of the majority of the cases.

scoder commented:
Thanks! Could you please also add the example from the question here?
https://stackoverflow.com/questions/16467479/normalizing-unicode
Maybe there are also some combining characters that do not go together with `a`? (Because that's what you used to determine the continuation characters in the generator script.)

One cool feature in the test runner is that it allows writing tests in `.py` (not `.pyx`) files that are then also executed in normal Python for comparison. If they require a specific minimum Python version (such as 3.0 in this case), you can add a tag like `# tag: pure3.0` at the top to disable them in pure Python 2.x. The compiled module will then still be tested in all Python versions. Thus, you can actually move the Cython specific parts of your tests into a separate `.pyx` test module and rename the main one to `.py`, so that we can get the confirmation that Cython behaves like Python for all those nice test cases.

That said, maybe an error case would also be helpful? Look in `tests/errors/` (with `# mode: error`), those tests can check compile errors, (hopefully) including invalid identifier characters.

scoder commented:
Note that error cases for the scanner are a bit annoying because each single failure is fatal. Testing multiple cases tends to require a separate test module for each of them, rather than just separate functions or even code lines.

da-woods commented:
I've added some more testing for normalized symbols including the a-acute example. It's in a slightly odd srctree structure because constructing examples from characters that look identical seemed better to do with escape sequences and string formatting. I've also added a few tests of error cases, although I don't think they're hugely interesting.

I don't think there's an issue with combining characters that don't go with "a":
1. If you look at the CPython source (https://github.com/python/cpython/blob/2a16eea71f56c2d8f38c295c8ce71a9a9a140aff/Objects/unicodeobject.c#L12200) it goes through character by character so the choice of first character doesn't affect whether the second is valid.
2. In PEP 3131 https://www.python.org/dev/peps/pep-3131/#id16 XID_start and XID_continue are described as properties of the character. They consider cases where 1 character normalizes to multiple characters and exclude if the multiple characters are invalid. When 2 characters normalizes to 1 doesn't affect the XIDs.

With that said the tests do now include combining characters that don't normalize to anything shorter (the normalization only standardizes the order of the combining characters)

da-woods commented:
I've added one further test. This required a bit of tweaking of the test runner to be able to match the output. Therefore it might be reasonable to decide that this test isn't worth the amount of change it needed...

scoder commented on runtests.py:
```diff
@@ -1002,10 +1002,18 @@ def source_files(self, workdir, module_name, file_list):
+            encoding = detect_opened_file_encoding(f, default='ISO-8859-1')
+        if encoding.lower() == 'ascii':
+            encoding = 'ISO-8859-1' # at least one test is based around a file tagged with ascii
+            # but with a character that can't be read with ascii. Therefore use different default
```
I think this is a very good test, but it's also a bit risky to play with the encoding here. This code was written to provide a pass-through, and anything we change here might make tests (even existing tests) silently test something other than they were written for. In some error test cases, the encoding declaration is also intentionally wrong, IIRC.

Instead, maybe we could decode only the error messages (see `ErrorWriter`), and not the entire test source file?

da-woods replied:
How about if I add an alternate way of specifying the errors so it first searches for `whatever_the_filename_is.errors`? I imagine most of the time it's still better to keep the errors with the code, but that seems more robust than trying to guess the source file encoding (which, as you say, can be intentionally wrong).

scoder commented:
I'd rather keep the complexity low for this. Let's assume that the errors can be decoded with the declared source code encoding. In almost all cases, they will be ASCII-only.

da-woods commented:
OK. I've done it the way you suggest - decoding in ErrorWriter is needed.
